### DISCUSSION TITLE
Minimal case study of using SYCL from cpp2

### DISCUSSION THREAD

**Discussion by MiroPalmu** [hsutter/cppfront#853] [Show and tell]
# Motivation

Cpp1 is the way to program GPUs in scientific computing with many different programming models.
One of the interesting models is SYCL which aims to accomplish this with standard cpp1 but then interpret some functions as device code and use special compilers to compile it to correct target.

[Intel describes](https://www.intel.com/content/www/us/en/developer/articles/technical/quick-guide-to-sycl-implementations.html) SYCL as:

> SYCL is a royalty-free open standard developed by the Khronos Group that allows developers to program heterogeneous architectures in standard C++. In addition, its programming model uses a single source, allowing both host and kernel code to be written in a single source file.

Newest spec can be found [here](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html).

Security goals of cpp2 might not be as desired in scientific computing as in other domains,
but simplification of language make it interesting for computational scientist,
who is more interested in just making algorithms work than in the technical details of the code.

(most of all I'm just interested to try cpp2)

# Goals of case study

- transform minimal cpp1 SYCL program to cpp2 and successfully compile it with [AdaptiveCpp](https://github.com/AdaptiveCpp/AdaptiveCpp)
- report found pain points.

# Cpp1 SYCL source

This is modified example from SYCL spec [section 3.2](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:anatomy)

```
#include <iostream>
#include <format>
#include <ranges>
#include <vector>
#include <sycl/sycl.hpp>

static constexpr auto N = std::size_t{ 1024 };

int main() {
    auto data = std::vector<int>(N);
    // Create a default queue to enqueue work to the default device
    auto queue = sycl::queue{ };

    {
        auto device_buffer = sycl::buffer{data};

        queue.submit([&](auto& handler) {
            const auto data_acc = sycl::accessor(device_buffer, handler, sycl::write_only, { sycl::no_init });

            handler.parallel_for(device_buffer.size(), [=](auto idx) {data_acc[idx] = idx; });
        });
    }

    // Print result
    for (const auto [i, x] : data | std::views::enumerate)
        std::cout << std::format("data[{}] = {}\n", i, x);
}
```

# Cpp2 SYCL source

cppfront commit `0b333f3` is able to transpile following code to form that is compile-able with AdaptiveCpp.

```
#include <sycl/sycl.hpp>

N: std::size_t == 1024;
main: () = {
    // data: std::vector<int> = (N); // Creates vector with one element of value N
    data := std::vector<int>(N);

    // Create a default queue to enqueue work to the default device
    queue: sycl::queue = (); 

    {
        device_buffer: sycl::buffer = (data);

        // Returns sycl::event which is ignored
        // handler is inout because accessor takes it as reference
        _ = queue.submit(:(inout handler) = {

            data_acc: const sycl::accessor = (device_buffer&$*, handler, sycl::write_only, sycl::property_list(sycl::no_init));

            handler.parallel_for(device_buffer&$*.size(), :(idx) data_acc$[idx] = idx;);
        });
    }

    // Print result
    for data | std::views::enumerate do (x)
          std::cout << "data[(x.get<0>())$] = (x.get<1>())$\n";
}
```

# Pain points

Keep in mind that these points might be due to authors lack of knowledge.

Order of points is arbitrary and does not indicate severity.

1) Commit `4bd0c04` force mutability to anonymous functions

    In SYCL kernel functions are callable objects (most often lambdas) which get executed on device (gpu, etc.).
    SYCL spec states that lambda kernel object can not be `mutable`.
    Stated commit adds `mutable` to anonymous functions, so this commit can not be used to write kernel functions as lambda.
    
    This is why previous commit was used.

2) Not being able to write default capture types

    SYCL defines command group scope as the callable object that gets passed to `queue.submit`,
    in which `sycl::buffer`s are needed as reference and in kernel scope `sycl::accessor`s
    are needed as values.

    This leads to that being able to specify default capture type is very convenient,
    which can be seen from [SYCL spec](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html) which uses `[&]` and `[=]` in every* example.

    In cpp2 we need to use `var&$*` and `var$` which is not as ergonomic as default captures.

3) Matching role based parameters with cpp1 parameters

    Seen in cpp1 code, lambda to `queue.submit(...)` has to take handler as reference.

    In all [SYCL spec](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html) examples
    all parameters to kernel functions are taken as values.
    (This might not be necessary due to `in` role working in cpp2 code. I haven't found definite answer.)

    It might not be obvious which roles to choose for the parameters in cpp2
    to match SYCL spec examples. This might be non-problem with enough experience with cpp2.

    In kernel function case `copy` parameter role might be desired to be able to match cpp1 examples which take parameter as value.

4) Having to write  `sycl::accessor(..., sycl::property_list(sycl::no_init))` and not `sycl::accessor(... , { sycl::no_init })`
5) Structured binding are not supported in `for` loop.

**Comment by JohelEGP** [comment]
For reference, these are the issues and comments related to the pain points:
1. <https://github.com/hsutter/cppfront/commit/4bd0c0438f2d3fa65d3e65a55b17c3a296bd8bc3#commitcomment-133307333> ("Commit `4bd0c04` force mutability to anonymous functions").
2. <https://github.com/hsutter/cppfront/issues/247#issuecomment-1449315482> ("Not being able to write default capture types").
3.
4. #542 (Having to write `sycl::accessor(..., sycl::property_list(sycl::no_init))` and not `sycl::accessor(... , { sycl::no_init })`).
5. #500 and linked issues ("Structured binding are not supported in `for` loop.").

**Comment by keryell** [comment]
Interesting to see some SYCL code written with cpp2!
